coremltools/models/utils.py

```diff
@@ -2251,4 +2252,199 @@ def _make_second_chunk_prog(prog: _mil.Program, op_idx: int) -> _mil.Program:
+        if feature_type == ArrayFeatureType.FLOAT16:
+            return fp16
+        if feature_type == ArrayFeatureType.FLOAT32:
+            return fp32
+        raise ValueError(f"invalid feature type: {feature_type}, supported only FLOAT16, FLOAT32")
+
+    def _sanitize_names(names: _Optional[_List[str]], desc_list: _Iterable, default: _List[str]) -> _List[str]:
+        if names is None:
+            names = default
+        return [x.name for x in desc_list if "*" in names or x.name in names]
+
+    def _eligible_feature_desc(feature_desc: FeatureDescription, names: _List[str], data_type: ArrayFeatureType) -> bool:
+        if feature_desc.name not in names:
+            _logger.debug(f"ignoring feature {feature_desc.name} as it's not in the list of required names {names}")
+            return False
+
+        feature_type = feature_desc.type.WhichOneof("Type")
+        if feature_type != "multiArrayType":
+            _logger.debug(f"ignoring output {feature_desc.name} (type: {feature_type})")
+            return False
+
+        feature_data_type = feature_desc.type.multiArrayType.dataType
+        if feature_data_type != data_type:
+            _logger.debug(f"ignoring output tensor {feature_desc.name} (data type: {feature_data_type})")
+            return False
+
+        return True
+
+    def _get_input_vars(var_name: str) -> _Iterable[_Tuple[_Optional[Function], _Optional[Var]]]:
+        for name in function_names:
+            func = prog.functions[name]
+            var = next(iter([v for k, v in func.inputs.items() if k == var_name]), None)
+            if var:
+                if func.opset_version < _ct.target.iOS16:
+                    _logger.warning(f"upgrading opset_version for function {func.name} to iOS16")
+                    func.opset_version = _ct.target.iOS16
+                yield func, var
+
+    def _get_output_vars(var_name: str) -> _Iterable[_Tuple[_Optional[Function], _Optional[Var]]]:
+        for name in function_names:
+            func = prog.functions[name]
+            var = next(iter([v for v in func.outputs if v.name == var_name]), None)
+            if var:
+                if func.opset_version < _ct.target.iOS16:
+                    _logger.warning(f"upgrading opset_version for function {func.name} to iOS16")
+                    func.opset_version = _ct.target.iOS16
+                yield func, var
+
+    def _cast_input_type(feature_desc: FeatureDescription, feature_var: Var, first_operation: Operation) -> None:
+        with first_operation.enclosing_block:
+            from_dtype_str = f"fp{from_dtype.get_bitwidth()}"
+            var_name = feature_desc.name + f"_to_{from_dtype_str}"
+            x = _mb.cast(x=feature_var, dtype=from_dtype_str, name=var_name, before_op=first_operation)
+            x.op.enclosing_block.replace_uses_of_var_after_op(
+                anchor_op=x.op,
+                old_var=feature_var,
+                new_var=x,
+                force_replace=True,
```

Comment:
Usually we only use force_replace for `constexpr_` related ops. Is there any specific reason to do force replacement here?

Reply:
No, not really. I've removed it and should be working fine even without the flag.